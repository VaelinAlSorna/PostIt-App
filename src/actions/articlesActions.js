import axios from 'axios';

if (process.env.NODE_ENV === 'production') {
    axios.defaults.baseURL= 'https://sleepy-shelf-83196.herokuapp.com'
} else {
    axios.defaults.baseURL = 'http://localhost:3000';
}

export const startSetArticles = () => {
    return async (dispatch) => {
        try {
            const response = await axios.get('/articles');
            dispatch(setArticles(response.data.allArticles));
            
        } catch (e) {
            if(e.message === 'Network Error'){
                dispatch(setNetworkError());
            } else {
                // it is here in else because I am testing this in test case, so in order to stop it from print out in terminal every single time I run test I put it in else statement, I am not going to test other thats why they are not in else
                console.log('Error in startSetArticles:', e); 
            }
        }
    };
};

// this action is called right above in the case that server is down it is handled by networkErrorReducer and also store in store in field networkError
export const setNetworkError = () => ({
    type: 'NETWORK_ERROR'
})

export const startAddArticle = (article) => {
    return async (dispatch, getState) => {
       try {
            const userToken = getState().auth.token
            const response = await axios.post('/article',{
                    "createdAt": article.createdAt,
                    "text": article.text,
                    "title": article.title
                },{
                headers: {'x-auth': userToken}
                }
            );

            dispatch(addArticle(response.data.article));
        } catch (e) {
            console.log('Error in startAddArticle:', e);
            if(e.message === 'Network Error'){
                dispatch(setNetworkError());
            }
        }
    };
};

export const startEditArticle = (articleId, article) => {
    return async (dispatch, getState) => {
       try {
            const userToken = getState().auth.token
            const response = await axios.patch(`/article/${articleId}`,{
                    "text": article.text,
                    "title": article.title,
                    "editedAt": article.createdAt //it's called createdAt but it represents time when it was updated, the name is createdAt because I use same form for creating and updating article
                },{
                headers: {'x-auth': userToken}
                }
            );
            dispatch(editArticle(articleId, response.data.updatedArticle));
        } catch (e) {
            console.log('Error in startEditArticle:', e);
            if(e.message === 'Network Error'){
                dispatch(setNetworkError());
            }
        }
    };
};

export const startRemoveArticle = (articleId) => {
    return async (dispatch, getState) => {
        try {
            const userToken = getState().auth.token;
            const response = await axios.delete(`/article/${articleId}`,{
                headers: {'x-auth': userToken}
                }
            );
            dispatch(removeArticle(response.data.removedArticle._id));
        } catch (e) {
            console.log('Error in startRemoveArticle', e);
            if(e.message === 'Network Error'){
                dispatch(setNetworkError());
            }
        }
    };
};

export const startCommentArticle = (articleId, comment) => {
    return async (dispatch, getState) => {
        try {
            const userToken = getState().auth.token;
            const response = await axios.post(`/article/${articleId}/comment`,{
                    text: comment.comment,
                    createdAt: comment.createdAt
                },{
                headers: {'x-auth': userToken}
                }
            );
            dispatch(commentArticle(articleId, response.data.comment));
        } catch (e) {
            console.log('Error in startCommentArticle', e);
            if(e.message === 'Network Error'){
                dispatch(setNetworkError());
            }
        }
    };
}

export const startRemoveComment = (articleId, commentId) => {
    return async (dispatch, getState) => {
        try {
            const userToken = getState().auth.token;
            const response = await axios.delete(`/article/${articleId}/comment/${commentId}`,{
                headers: {'x-auth': userToken}
            });
            dispatch(removeComment(articleId, commentId));
        } catch (e) {
            console.log('Error in startRemoveComment', e);
            if(e.message === 'Network Error'){
                dispatch(setNetworkError());
            }
        }
    };
};

export const startAddLike = (articleId ,userId) => {
    return async (dispatch, getState) => {
        try {
            const userToken = getState().auth.token;
            await axios.post(`/article/${articleId}/like`,{},{
                headers: {'x-auth': userToken}
            });
            dispatch(addLike( articleId, userId));
        } catch (e) {
            console.log('Error in startAddLike', e);
            if(e.message === 'Network Error'){
                dispatch(setNetworkError());
            }
        }
    };
};

export const startRemoveLike = (articleId ,userId) => {
    return async (dispatch, getState) => {
        try {
            const userToken = getState().auth.token;
            await axios.delete(`/article/${articleId}/like` ,{
                headers: {'x-auth': userToken}
            });
            dispatch(removeLike( articleId, userId));
        } catch (e) {
            console.log('Error in startRemoveLike', e);
            if(e.message === 'Network Error'){
                dispatch(setNetworkError());
            }
        }
    };
};

export const setArticles = (articles) => ({
    type: 'SET_ARTICLES',
    articles
});

export const addArticle = (article) => ({
    type: 'ADD_ARTICLE',
    article
});

export const editArticle = (id, updatedArticle) => ({
    type: 'EDIT_ARTICLE',
    id, 
    updatedArticle
});

export const removeArticle = (id) => ({
    type: 'REMOVE_ARTICLE',
    id
}); 

export const commentArticle = (id, comment) => ({
    type: 'COMMENT_ARTICLE',
    id,
    comment
});

export const removeComment = (id, commentId) => ({
    type: 'REMOVE_COMMENT',
    id,
    commentId
});

export const addLike = ( id,userId) => ({
    type: 'ADD_LIKE',
    id,
    userId
});

export const removeLike = ( id,userId) => ({
    type: 'REMOVE_LIKE',
    id,
    userId
});