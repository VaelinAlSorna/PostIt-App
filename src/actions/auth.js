import axios from 'axios';

export const SignUp = (user) => ({
    type: 'SIGN_UP',
    user
});

export const startSignUp = (userData) => {
    return async (dispatch) => {
        try {
            const response = await axios({
                method: 'post',
                url: 'http://localhost:3000/user',
                data: {
                    "name": userData.name,
                    "password": userData.password
                }
            })
            const newUser = {
            id: response.data.user._id,
            name: response.data.user.name,
            token: response.headers["x-auth"]
            };
            dispatch(SignUp(newUser))
        } catch (e) {
            console.log('Error startSignUp', e);
        }
    };
};
