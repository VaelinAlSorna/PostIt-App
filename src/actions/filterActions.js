
export const setYourArticlesFilter = () => ({
    type: 'SET_YOUR_ARTICLES_FILTER'

});

export const removeYourArticlesFilter = () => ({
    type: 'REMOVE_YOUR_ARTICLES_FILTER'
});

export const setTextFilter = (text) => ({
    type: 'SET_TEXT_FILTER',
    text
});