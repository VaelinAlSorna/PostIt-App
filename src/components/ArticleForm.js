import React from 'react';
import moment from 'moment';

class ArticleForm extends React.Component {
    state = {
        title: '',
        text: '',
        createdAt: moment().format('D. M. Y')
    };
    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState(() => ({
            title
        }))
    }
    onTextChange = (e) => {
        const text = e.target.value;
        this.setState(() => ({
            text
        }))
    }
    onSubmit = (e) => {
        e.preventDefault();
        this.props.onSubmit(this.state)
    }
    render () {
       return ( 
        <div> 
            <form onSubmit={this.onSubmit}> 
                <input 
                    onChange={this.onTitleChange} 
                    placeholder="Title" 
                    type="text" 
                    value={this.state.title}
                />
                <textarea 
                    onChange={this.onTextChange}
                    placeholder="Add text for your article"
                    value={this.state.text}
                >
                </textarea>
                <input type="submit" value="Submit" />
            </form>
        </div>
       );
    }
}

export default ArticleForm;