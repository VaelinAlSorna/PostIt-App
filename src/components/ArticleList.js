import React from 'react';
import { connect } from 'react-redux';
import ArticleListItem from './ArticleListItem';
import ArticleListFilter from './ArticleListFilter';
import yourArticlesFilter from '../selectors/yourArticlesFilter';

export const ArticleList = (props) => (
    <div className="main-content responsive-scroll">
        <ArticleListFilter />
        <ol className="list">
            {
            !!props.networkError
                ? <p className="server-error"> We are sorry ,but server is not responding ,please try to reconnect later.</p>   
                : (props.articles.length === 0) 
                    ? <p className="no-articles"> No articles fulfill these search conditions. </p>
                    : props.articles.map((article) => <ArticleListItem key={article._id} {...article} />)
            }
        </ol>
    </div>
);
 

const mapStateToProps = (state) => ({
    articles: yourArticlesFilter(state.articles, state.auth.id, state.filter),
    networkError: state.networkError
});

export default connect(mapStateToProps)(ArticleList);