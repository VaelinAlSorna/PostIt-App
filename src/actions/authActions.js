import axios from 'axios';
import { setNetworkError } from './articlesActions';

axios.defaults.baseURL = 'http://localhost:3000';

//Signing Up
export const startSignUp = (userData) => {
    return async (dispatch) => {
        try {
            const response = await axios.post( '/user', {
                "name": userData.name,
                "password": userData.password
            });
            const newUser = {
            // user id is not stored _id (thah way I store it for articles and comments)
            id: response.data.user._id,
            name: response.data.user.name,
            token: response.headers["x-auth"]
            };

            sessionStorage.setItem('token', newUser.token);
            dispatch(storeUser(newUser))
        } catch (e) {
            console.log('Error in startSignUp:', e);
            if(e.message === 'Network Error'){
                dispatch(setNetworkError());
            }
            //login property here helps me determine on which form I should render error
            dispatch(apiError({error:'User with this name already exists.' ,login: false}));
        }
    };
};

//Loging In
export const startLogin = (userData) => {
    return async (dispatch) => {
       try { 
            const response = await axios.post( '/user/login', {
                "name": userData.name,
                "password": userData.password
            });
            const newUser = {
            id: response.data.user._id,
            name: response.data.user.name,
            token: response.headers["x-auth"]
            };

            sessionStorage.setItem('token', newUser.token);
            dispatch(storeUser(newUser))
        } catch (e) {
            console.log('Error in startLogin:', e);
            if(e.message === 'Network Error'){
                dispatch(setNetworkError());
            }
            //login property here helps me determine on which form I should render error
            dispatch(apiError({error:'Wrong name or password.', login: true}));
        }
    };
};

//get user on page refresh
export const getUser = (userToken) => {
    return async (dispatch) => {
        try { 
             const response = await axios({
                method: 'get',
                url: '/user/me',
                headers: {'x-auth': userToken}
            });
             const newUser = {
             id: response.data.user._id,
             name: response.data.user.name,
             token: userToken
             };
 
             dispatch(storeUser(newUser))
         } catch (e) {
             //login property here helps me determine on which form I should render error
             console.log('Error in getUser:', e);
            if(e.message === 'Network Error'){
                dispatch(setNetworkError());
            }
         }
     };
}

//startLogout
export const startLogout = (userToken) => {
    return async (dispatch) => {
    try {
        const response = await axios({
             method: 'delete',
             url: '/user/logout',
             headers: {'x-auth': userToken}
         });
         sessionStorage.removeItem('token');
         dispatch(removeUser());
    } catch (e) {
        console.log('Error in startLogout:', e);
        if(e.message === 'Network Error'){
            dispatch(setNetworkError());
        }
    }
    };
};

//Storing User afer startSignUp or startLogin
export const storeUser = (user) => ({
    type: 'STORE_USER',
    user
});

//Handling 400 bad request sent back from server when wrong data is passed in
export const apiError = (error) => ({
    type: 'API_ERROR',
    error
});

export const removeUser = () => ({
    type: 'REMOVE_USER'
});