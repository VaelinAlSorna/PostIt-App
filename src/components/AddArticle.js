import React from 'react';
import { connect } from 'react-redux'
import ArticleForm from './ArticleForm';
import { addArticle } from '../actions/articlesActions'


class AddArticle extends React.Component {
    onSubmit = (article) => {
        this.props.dispatch(addArticle(article))
        this.props.history.push('/');
    }
    render () {
        return ( 
        <div>
            <ArticleForm onSubmit={this.onSubmit} />
        </div>
        );
    }
}

export default connect(undefined)(AddArticle);

