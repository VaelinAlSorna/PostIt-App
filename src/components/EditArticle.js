import React from 'react';
import { connect } from 'react-redux';
import ArticleForm from './ArticleForm';
import { editArticle, removeArticle } from '../actions/articlesActions';

export class EditArticle extends React.Component {
    onSubmit = (article) => {
        this.props.editArticle(this.props.article._id, article);
        this.props.history.push('/');
    }
    onRemoveClick = () => {
        this.props.removeArticle(this.props.article._id);
        this.props.history.push('/');
    }
    render () {
        return (
            <div>
                <h2>Editing Article</h2>
                <ArticleForm onSubmit={this.onSubmit} article={this.props.article}/>
                <button onClick={this.onRemoveClick}>Remove Article</button>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => ({
    article: state.articles.find((article) => article._id === props.match.params.id )
});

const mapDispatchToProps = (dispatch) => ({
    editArticle: (id, article) => dispatch(editArticle(id, article)),
    removeArticle: (id) => dispatch(removeArticle(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditArticle);