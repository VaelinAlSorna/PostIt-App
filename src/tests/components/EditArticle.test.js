import React from 'react';
import { shallow } from 'enzyme';
import { EditArticle } from '../../components/EditArticle';
import articles from '../fixtures/articlesFixtures';
import { resolve } from 'url';
import { rejects } from 'assert';
import { editArticle } from '../../actions/articlesActions';

let wrapper, startEditArticle, startRemoveArticle, history;

beforeEach(() => {
    startEditArticle = jest.fn(() => { return Promise.resolve(); });
    startRemoveArticle = jest.fn(() => { return Promise.resolve(); });
    history = { push: jest.fn() };
    wrapper = shallow(<EditArticle 
                                article={articles[0]}
                                startEditArticle={startEditArticle}
                                history={history}
                                startRemoveArticle={startRemoveArticle}
                            />
                        );
});

test('should render EditArticle correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle startEditArticle', async () => {
    const editedArticle = {title: 'edited title', text: 'edited text', createdAt: '2.7.2018' }
    await wrapper.find('ArticleForm').prop('onSubmit')(editedArticle);
    expect(startEditArticle).toHaveBeenLastCalledWith(articles[0]._id, editedArticle);
    expect(history.push).toHaveBeenLastCalledWith('/');
});

test('should handle startRemoveArticle', async () => {
    wrapper.find('button').simulate('click');
    await expect(startRemoveArticle).toHaveBeenLastCalledWith(articles[0]._id);
    expect(history.push).toHaveBeenLastCalledWith('/');
});