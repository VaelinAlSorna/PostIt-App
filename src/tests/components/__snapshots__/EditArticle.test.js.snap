// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditArticle correctly 1`] = `
<div
  className="main-content"
>
  <div>
    <h2
      className="add-edit-part u-margin-bottom-medium"
    >
      Editing Article
    </h2>
    <ArticleForm
      article={
        Object {
          "_creatorId": "343",
          "_id": "1",
          "comments": Array [
            Object {
              "_id": "1a",
              "comment": "Something to test?",
              "createdAt": "1.1.1970",
            },
          ],
          "createdAt": "1.1.1970",
          "likes": Array [
            "343",
          ],
          "text": "Political info",
          "title": "Breaking news",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className=" add-edit-part__remove btn btn--white"
      onClick={[Function]}
    >
      Remove Article
    </button>
  </div>
</div>
`;

exports[`should render error when article does note exist 1`] = `
<div
  className="main-content"
>
  <p>
    You haven't created this article
  </p>
</div>
`;

exports[`should render error when articleId does not match with userId 1`] = `
<div
  className="main-content"
>
  <p>
    You haven't created this article
  </p>
</div>
`;
